File: frontend/src/api/sponsors.js
import { apiClient, USE_MOCK } from './client'
import { delay, paginate } from './mockUtils'
import { mapMediaRef } from '../utils/media'

// No dedicated Sponsors mock dataset exists yet (this module is new) — a
// small self-contained fixture keeps mock mode functional without
// growing the shared mock/admin.js file for a shape nothing else reads.
const MOCK_SPONSORS = [
  {
    id: 1,
    campaign_name: 'Kaziwave Leadership Series',
    organization: { id: 1, slug: 'kaziwave', name: 'Kaziwave', logo: null },
    partnership: null,
    sponsorship_type: 'Content Sponsor',
    status: 'active',
    starts_at: '2026-01-01',
    ends_at: '2026-12-31',
    logo: null,
    creative: null,
    sponsor_url: 'https://kaziwave.example.com',
    cta_label: 'Visit Sponsor',
    disclosure_label: 'Sponsored by',
    public_visible: true,
    is_exclusive: false,
    exclusivity_notes: '',
    tier: 'Gold',
    estimated_value: 15000,
    currency: 'USD',
    commercial_notes: '',
    internal_notes: '',
    internal_reference: '',
    public_name_override: '',
    public_description: 'Supporting our Leadership Series through 2026.',
    placements: [{ id: 1, placement_key: 'homepage_featured', position: 0, starts_at: null, ends_at: null, active: true }],
    created_at: '2026-01-01T00:00:00Z',
    updated_at: '2026-01-01T00:00:00Z',
  },
]

function mapPlacement(p) {
  if (!p) return null
  return {
    id: p.id,
    placementKey: p.placement_key,
    position: p.position,
    startsAt: p.starts_at,
    endsAt: p.ends_at,
    active: p.active,
  }
}

export function mapSponsor(s) {
  if (!s) return null
  return {
    id: s.id,
    campaignName: s.campaign_name,
    organization: s.organization ? { id: s.organization.id, slug: s.organization.slug, name: s.organization.name, logo: mapMediaRef(s.organization.logo) } : null,
    organizationSlug: s.organization?.slug || null,
    partnership: s.partnership ? { id: s.partnership.id, company: s.partnership.company, subject: s.partnership.subject, status: s.partnership.status } : null,
    partnershipId: s.partnership?.id || null,
    internalReference: s.internal_reference || '',
    publicNameOverride: s.public_name_override || '',
    publicDescription: s.public_description || '',
    sponsorshipType: s.sponsorship_type || '',
    startsAt: s.starts_at || '',
    endsAt: s.ends_at || '',
    logo: mapMediaRef(s.logo),
    creative: mapMediaRef(s.creative),
    sponsorUrl: s.sponsor_url || '',
    ctaLabel: s.cta_label || '',
    disclosureLabel: s.disclosure_label || 'Sponsored by',
    status: s.status || 'draft',
    publicVisible: !!s.public_visible,
    isExclusive: !!s.is_exclusive,
    exclusivityNotes: s.exclusivity_notes || '',
    tier: s.tier || '',
    estimatedValue: s.estimated_value ?? null,
    currency: s.currency || '',
    commercialNotes: s.commercial_notes || '',
    internalNotes: s.internal_notes || '',
    placements: Array.isArray(s.placements) ? s.placements.map(mapPlacement) : [],
    createdAt: s.created_at || null,
    updatedAt: s.updated_at || null,
  }
}

function buildSponsorPayload(payload) {
  return {
    campaignName: payload.campaignName,
    organizationSlug: payload.organizationSlug || undefined,
    partnershipId: payload.partnershipId || undefined,
    internalReference: payload.internalReference || undefined,
    publicNameOverride: payload.publicNameOverride || undefined,
    publicDescription: payload.publicDescription || undefined,
    sponsorshipType: payload.sponsorshipType || undefined,
    startsAt: payload.startsAt || undefined,
    endsAt: payload.endsAt || undefined,
    logoMediaId: payload.logo?.id || undefined,
    creativeMediaId: payload.creative?.id || undefined,
    sponsorUrl: payload.sponsorUrl || undefined,
    ctaLabel: payload.ctaLabel || undefined,
    disclosureLabel: payload.disclosureLabel || undefined,
    publicVisible: payload.publicVisible,
    isExclusive: payload.isExclusive,
    exclusivityNotes: payload.exclusivityNotes || undefined,
    tier: payload.tier || undefined,
    estimatedValue: payload.estimatedValue === '' || payload.estimatedValue == null ? undefined : Number(payload.estimatedValue),
    currency: payload.currency || undefined,
    commercialNotes: payload.commercialNotes || undefined,
    internalNotes: payload.internalNotes || undefined,
  }
}

// ---------------------------------------------------------------------------
// Admin
// ---------------------------------------------------------------------------

export async function fetchSponsors(params = {}) {
  if (!USE_MOCK) {
    const { data } = await apiClient.get('/sponsors/', { params })
    return { ...data, items: data.items.map(mapSponsor) }
  }
  let results = MOCK_SPONSORS
  if (params.status) results = results.filter((s) => s.status === params.status)
  if (params.q) {
    const q = params.q.toLowerCase()
    results = results.filter((s) => s.campaign_name.toLowerCase().includes(q))
  }
  return delay(paginate(results.map(mapSponsor), params))
}

export async function fetchSponsor(id) {
  if (!USE_MOCK) {
    const { data } = await apiClient.get(`/sponsors/${id}`)
    return mapSponsor(data)
  }
  return delay(mapSponsor(MOCK_SPONSORS.find((s) => String(s.id) === String(id))))
}

export async function createSponsor(payload) {
  const { data } = await apiClient.post('/sponsors/', buildSponsorPayload(payload))
  return mapSponsor(data)
}

export async function updateSponsor(id, payload) {
  const { data } = await apiClient.patch(`/sponsors/${id}`, buildSponsorPayload(payload))
  return mapSponsor(data)
}

export async function updateSponsorStatus(id, status) {
  const { data } = await apiClient.patch(`/sponsors/${id}/status`, { status })
  return mapSponsor(data)
}

export async function deleteSponsor(id) {
  await apiClient.delete(`/sponsors/${id}`)
}

export async function addSponsorPlacement(id, payload) {
  const { data } = await apiClient.post(`/sponsors/${id}/placements`, {
    placementKey: payload.placementKey,
    position: payload.position ?? 0,
    startsAt: payload.startsAt || undefined,
    endsAt: payload.endsAt || undefined,
    active: payload.active,
  })
  return mapSponsor(data)
}

export async function updateSponsorPlacement(id, placementId, payload) {
  const { data } = await apiClient.patch(`/sponsors/${id}/placements/${placementId}`, {
    position: payload.position,
    startsAt: payload.startsAt || undefined,
    endsAt: payload.endsAt || undefined,
    active: payload.active,
  })
  return mapSponsor(data)
}

export async function removeSponsorPlacement(id, placementId) {
  const { data } = await apiClient.delete(`/sponsors/${id}/placements/${placementId}`)
  return mapSponsor(data)
}

export async function fetchSponsorHistory(id) {
  if (!USE_MOCK) {
    const { data } = await apiClient.get(`/sponsors/${id}/history`)
    return data
  }
  return delay([])
}

export async function fetchSponsorAnalytics(id) {
  if (!USE_MOCK) {
    const { data } = await apiClient.get(`/sponsors/${id}/analytics`)
    return data
  }
  return delay({ impressions: 0, clicks: 0, clickThroughRate: null })
}

// ---------------------------------------------------------------------------
// Public
// ---------------------------------------------------------------------------

function mapPublicSponsor(s) {
  return {
    id: s.id,
    campaignName: s.campaignName,
    publicName: s.publicName,
    publicDescription: s.publicDescription,
    organization: s.organization,
    logo: mapMediaRef(s.logo),
    creative: mapMediaRef(s.creative),
    disclosureLabel: s.disclosureLabel,
    sponsorUrl: s.sponsorUrl,
    ctaLabel: s.ctaLabel,
    placementKey: s.placementKey,
  }
}

export async function fetchPublicSponsorPlacements(placementKey) {
  if (!USE_MOCK) {
    const { data } = await apiClient.get('/sponsors/public', { params: { placement: placementKey } })
    return data.map(mapPublicSponsor)
  }
  return delay(
    MOCK_SPONSORS.filter((s) => s.status === 'active' && s.public_visible && s.placements.some((p) => p.placement_key === placementKey)).map((s) =>
      mapPublicSponsor({
        id: s.id,
        campaignName: s.campaign_name,
        publicName: s.public_name_override || s.organization?.name,
        publicDescription: s.public_description,
        organization: s.organization,
        logo: s.logo,
        creative: s.creative,
        disclosureLabel: s.disclosure_label,
        sponsorUrl: s.sponsor_url,
        ctaLabel: s.cta_label,
        placementKey,
      }),
    ),
  )
}

File: backend/app/api/v1/sponsors.py
from datetime import date

from flask import Blueprint, request
from flask_jwt_extended import current_user
from flask_restful import Api, Resource

from app.auth.decorators import permission_required
from app.extensions import db
from app.models.article import Article
from app.models.audit import AuditLog
from app.models.analytics import AnalyticsEvent
from app.models.commerce import PartnershipInquiry, Sponsor, SponsorPlacement, SPONSOR_PLACEMENT_KEYS
from app.models.opportunity import Job
from app.models.people import Organization
from app.schemas.commerce import (
    SponsorAdminUpdateSchema,
    SponsorInputSchema,
    SponsorPlacementInputSchema,
    SponsorPlacementSchema,
    SponsorSchema,
    SponsorStatusInputSchema,
)
from app.schemas.media import MediaSchema
from app.services.audit import log_action
from app.utils.filtering import apply_equality_filters, apply_search
from app.utils.pagination import paginate
from app.utils.responses import ApiError, success_response

sponsors_bp = Blueprint("sponsors", __name__)
api = Api(sponsors_bp)

sponsor_schema = SponsorSchema()
placement_schema = SponsorPlacementSchema()


def _resolve_organization(slug):
    organization = Organization.query.filter_by(slug=slug).first()
    if organization is None:
        raise ApiError(f'Organization "{slug}" not found.', 404, code="not_found")
    return organization


def _resolve_partnership(partnership_id):
    if not partnership_id:
        return None
    partnership = db.session.get(PartnershipInquiry, partnership_id)
    if partnership is None:
        raise ApiError("Linked partnership not found.", 404, code="not_found")
    return partnership


def _get_sponsor_or_404(sponsor_id):
    sponsor = db.session.get(Sponsor, sponsor_id)
    if sponsor is None:
        raise ApiError("Sponsor not found.", 404, code="not_found")
    return sponsor


def _build_query():
    query = Sponsor.query.order_by(Sponsor.updated_at.desc())
    query = apply_equality_filters(query, Sponsor, request.args, ["status", "sponsorship_type"])
    query = apply_search(query, Sponsor, request.args, ["campaign_name", "internal_reference"])

    organization_slug = request.args.get("organization")
    if organization_slug:
        query = query.join(Organization).filter(Organization.slug == organization_slug)

    placement = request.args.get("placement")
    if placement:
        query = query.join(SponsorPlacement).filter(SponsorPlacement.placement_key == placement)

    availability = request.args.get("availability")  # "active" | "expired"
    today = date.today()
    if availability == "active":
        query = query.filter(db.or_(Sponsor.ends_at.is_(None), Sponsor.ends_at >= today))
    elif availability == "expired":
        query = query.filter(Sponsor.ends_at.isnot(None), Sponsor.ends_at < today)

    date_from = request.args.get("date_from")
    if date_from:
        query = query.filter(db.or_(Sponsor.ends_at.is_(None), Sponsor.ends_at >= date_from))
    date_to = request.args.get("date_to")
    if date_to:
        query = query.filter(db.or_(Sponsor.starts_at.is_(None), Sponsor.starts_at <= date_to))

    return query.distinct()


def _apply_fields(sponsor, data):
    """Applies every SponsorInputSchema/SponsorAdminUpdateSchema field
    except organization/partnership (resolved separately) and status
    (dedicated action only).
    """
    for field in (
        "campaign_name", "internal_reference", "public_name_override", "public_description",
        "sponsorship_type", "starts_at", "ends_at", "logo_media_id", "creative_media_id",
        "sponsor_url", "cta_label", "disclosure_label", "public_visible", "is_exclusive",
        "exclusivity_notes", "tier", "estimated_value", "currency", "commercial_notes", "internal_notes",
    ):
        if field in data:
            setattr(sponsor, field, data[field])

    if "organization_slug" in data and data["organization_slug"]:
        sponsor.organization = _resolve_organization(data["organization_slug"])
    if "partnership_id" in data:
        sponsor.partnership = _resolve_partnership(data["partnership_id"])


class SponsorListResource(Resource):
    @permission_required("partnerships.manage")
    def get(self):
        result = paginate(_build_query(), sponsor_schema)
        return success_response(result["items"], meta=result["meta"])

    @permission_required("partnerships.manage")
    def post(self):
        data = SponsorInputSchema().load(request.get_json(silent=True) or {})
        organization = _resolve_organization(data.pop("organization_slug"))
        partnership = _resolve_partnership(data.pop("partnership_id", None))
        sponsor = Sponsor(organization=organization, partnership=partnership, **data)
        db.session.add(sponsor)
        db.session.commit()
        log_action(current_user, "sponsor.create", "Sponsor", sponsor.id)
        return success_response(sponsor_schema.dump(sponsor), status=201)


class SponsorDetailResource(Resource):
    @permission_required("partnerships.manage")
    def get(self, sponsor_id):
        return success_response(sponsor_schema.dump(_get_sponsor_or_404(sponsor_id)))

    @permission_required("partnerships.manage")
    def patch(self, sponsor_id):
        sponsor = _get_sponsor_or_404(sponsor_id)
        data = SponsorAdminUpdateSchema().load(request.get_json(silent=True) or {})
        _apply_fields(sponsor, data)
        try:
            db.session.commit()
        except Exception as exc:
            db.session.rollback()
            raise ApiError("Couldn't save — check that the end date isn't before the start date.", 422, code="validation_error") from exc
        log_action(current_user, "sponsor.update", "Sponsor", sponsor.id)
        return success_response(sponsor_schema.dump(sponsor))

    @permission_required("partnerships.manage")
    def delete(self, sponsor_id):
        sponsor = _get_sponsor_or_404(sponsor_id)
        if sponsor.status != "draft":
            raise ApiError(
                "Only Draft sponsors can be deleted. Archive established sponsors instead.",
                409, code="delete_restricted",
            )
        referenced = (
            Job.query.filter_by(sponsor_id=sponsor.id).count()
            or Article.query.filter_by(sponsor_id=sponsor.id).count()
        )
        if referenced:
            raise ApiError(
                "This sponsor is referenced by other content and can't be deleted. Archive it instead.",
                409, code="delete_restricted",
            )
        db.session.delete(sponsor)
        db.session.commit()
        log_action(current_user, "sponsor.delete", "Sponsor", sponsor_id)
        return success_response(None, status=204)


class SponsorStatusResource(Resource):
    @permission_required("partnerships.manage")
    def patch(self, sponsor_id):
        sponsor = _get_sponsor_or_404(sponsor_id)
        data = SponsorStatusInputSchema().load(request.get_json(silent=True) or {})
        previous_status = sponsor.status
        sponsor.status = data["status"]
        db.session.commit()
        log_action(
            current_user, "sponsor.status_change", "Sponsor", sponsor.id,
            changes={"from": previous_status, "to": sponsor.status},
        )
        return success_response(sponsor_schema.dump(sponsor))


class SponsorPlacementListResource(Resource):
    @permission_required("partnerships.manage")
    def post(self, sponsor_id):
        sponsor = _get_sponsor_or_404(sponsor_id)
        data = SponsorPlacementInputSchema().load(request.get_json(silent=True) or {})
        placement = SponsorPlacement(sponsor=sponsor, **data)
        db.session.add(placement)
        db.session.commit()
        log_action(
            current_user, "sponsor.placement_add", "Sponsor", sponsor.id,
            changes={"placementKey": placement.placement_key},
        )
        return success_response(sponsor_schema.dump(sponsor), status=201)


class SponsorPlacementDetailResource(Resource):
    @permission_required("partnerships.manage")
    def patch(self, sponsor_id, placement_id):
        sponsor = _get_sponsor_or_404(sponsor_id)
        placement = SponsorPlacement.query.filter_by(id=placement_id, sponsor_id=sponsor.id).first()
        if placement is None:
            raise ApiError("Placement not found.", 404, code="not_found")
        data = SponsorPlacementInputSchema(partial=True).load(request.get_json(silent=True) or {})
        for field, value in data.items():
            setattr(placement, field, value)
        db.session.commit()
        log_action(current_user, "sponsor.placement_update", "Sponsor", sponsor.id, changes={"placementId": placement.id})
        return success_response(sponsor_schema.dump(sponsor))

    @permission_required("partnerships.manage")
    def delete(self, sponsor_id, placement_id):
        sponsor = _get_sponsor_or_404(sponsor_id)
        placement = SponsorPlacement.query.filter_by(id=placement_id, sponsor_id=sponsor.id).first()
        if placement is None:
            raise ApiError("Placement not found.", 404, code="not_found")
        placement_key = placement.placement_key
        db.session.delete(placement)
        db.session.commit()
        log_action(
            current_user, "sponsor.placement_remove", "Sponsor", sponsor.id, changes={"placementKey": placement_key}
        )
        return success_response(sponsor_schema.dump(sponsor))


class SponsorHistoryResource(Resource):
    @permission_required("partnerships.manage")
    def get(self, sponsor_id):
        _get_sponsor_or_404(sponsor_id)
        entries = (
            AuditLog.query.filter_by(entity_type="Sponsor", entity_id=str(sponsor_id))
            .order_by(AuditLog.created_at.desc())
            .all()
        )
        return success_response(
            [
                {
                    "id": e.id,
                    "action": e.action,
                    "changes": e.changes,
                    "user": e.user.full_name if e.user else None,
                    "createdAt": e.created_at.isoformat() if e.created_at else None,
                }
                for e in entries
            ]
        )


class SponsorAnalyticsResource(Resource):
    """Real counts only — no fabricated metrics. CTR is omitted (not
    zeroed) when there are no impressions to divide by.
    """

    @permission_required("partnerships.manage")
    def get(self, sponsor_id):
        _get_sponsor_or_404(sponsor_id)
        rows = (
            db.session.query(AnalyticsEvent.event_name, db.func.count(AnalyticsEvent.id))
            .filter(AnalyticsEvent.entity_type == "Sponsor", AnalyticsEvent.entity_id == str(sponsor_id))
            .group_by(AnalyticsEvent.event_name)
            .all()
        )
        counts = {name: count for name, count in rows}
        impressions = counts.get("sponsor_impression", 0)
        clicks = counts.get("sponsor_click", 0)
        ctr = round(clicks / impressions * 100, 2) if impressions else None
        return success_response({"impressions": impressions, "clicks": clicks, "clickThroughRate": ctr})


def build_public_sponsor_payload(sponsor, placement):
    logo = sponsor.logo or (sponsor.organization.logo if sponsor.organization else None)
    return {
        "id": sponsor.id,
        "campaignName": sponsor.campaign_name,
        "publicName": sponsor.resolved_public_name,
        "publicDescription": sponsor.public_description,
        "organization": (
            {"slug": sponsor.organization.slug, "name": sponsor.organization.name}
            if sponsor.organization else None
        ),
        "logo": MediaSchema().dump(logo) if logo else None,
        "creative": MediaSchema().dump(sponsor.creative) if sponsor.creative else None,
        "disclosureLabel": sponsor.disclosure_label,
        "sponsorUrl": sponsor.sponsor_url,
        "ctaLabel": sponsor.cta_label,
        "placementKey": placement.placement_key,
        "position": placement.position,
    }


class PublicSponsorPlacementResource(Resource):
    def get(self):
        placement_key = request.args.get("placement")
        if not placement_key or placement_key not in SPONSOR_PLACEMENT_KEYS:
            raise ApiError("A valid placement is required.", 422, code="validation_error")

        today = date.today()
        rows = (
            db.session.query(Sponsor, SponsorPlacement)
            .join(SponsorPlacement, SponsorPlacement.sponsor_id == Sponsor.id)
            .filter(
                SponsorPlacement.placement_key == placement_key,
                SponsorPlacement.active.is_(True),
                Sponsor.status == "active",
                Sponsor.public_visible.is_(True),
                db.or_(Sponsor.starts_at.is_(None), Sponsor.starts_at <= today),
                db.or_(Sponsor.ends_at.is_(None), Sponsor.ends_at >= today),
                db.or_(SponsorPlacement.starts_at.is_(None), SponsorPlacement.starts_at <= today),
                db.or_(SponsorPlacement.ends_at.is_(None), SponsorPlacement.ends_at >= today),
            )
            .order_by(SponsorPlacement.position)
            .all()
        )
        return success_response([build_public_sponsor_payload(sponsor, placement) for sponsor, placement in rows])


api.add_resource(SponsorListResource, "/")
api.add_resource(PublicSponsorPlacementResource, "/public")
api.add_resource(SponsorDetailResource, "/<int:sponsor_id>")
api.add_resource(SponsorStatusResource, "/<int:sponsor_id>/status")
api.add_resource(SponsorPlacementListResource, "/<int:sponsor_id>/placements")
api.add_resource(SponsorPlacementDetailResource, "/<int:sponsor_id>/placements/<int:placement_id>")
api.add_resource(SponsorHistoryResource, "/<int:sponsor_id>/history")
api.add_resource(SponsorAnalyticsResource, "/<int:sponsor_id>/analytics")
